src/plugins/topology/checkDock.cpp

```diff
@@ -376,22 +376,20 @@ void checkDock::runTests( ValidateType type )
     for ( it = errors.begin(); it != errors.end(); ++it )
     {
       TopolError *te = *it;
-      te->conflict();
-
       const QgsSettings settings;
-      if ( te->conflict().type() == Qgis::GeometryType::Polygon )
-      {
-        rb = new QgsRubberBand( qgsInterface->mapCanvas(), Qgis::GeometryType::Polygon );
-      }
-      else
+      const QgsGeometry geom = te->conflict();
+      QVector<QgsGeometry> geoms = ( geom.wkbType() == Qgis::WkbType::GeometryCollection ?
+                                     geom.asGeometryCollection() : QVector<QgsGeometry>() << geom );
+
+      for ( const QgsGeometry &g : std::as_const( geoms ) )
```

Comment:
How about using:

```
        for ( auto it = geom.const_parts_begin(); it != geom.const_parts_end(); ++it )
        {
            ...
        }
```
instead? It's more efficient vs calling asGeometryCollection

Reply:
Since we need to give a QgsGeometry to QgsRubberBand, that would require to create a QgsGeometry from a const cloned QgsAbstractGometry, which will do the same thing than asGeometryCollection. 

Plus, we would create one rubberband for each part of multi part geom though QgsRubberBand knows how to deal with multipart.

On the other side, we spare only the list allocation, so I would say that's better this way